rewrite-java/src/main/java/org/openrewrite/java/BasicMethodMatcher.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java;
+
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+/**
+ * The most basic version of a {@link MethodMatcher} that allows implementers to craft custom matching logic.
+ */
+@FunctionalInterface
+public interface BasicMethodMatcher {
+
+    /**
+     * Whether the method invocation or constructor matches the criteria of this matcher.
+     *
+     * @param type The type of the method invocation or constructor.
+     * @return True if the invocation or constructor matches the criteria of this matcher.
+     */
+    boolean matches(@Nullable JavaType.Method type);
+
+    default boolean matches(@Nullable J.NewClass constructor) {
+        if (constructor == null) {
+            return false;
+        }
+        return matches(constructor.getConstructorType());
+    }
+
+    default boolean matches(@Nullable J.MethodInvocation method) {
+        if (method == null) {
+            return false;
+        }
+        return matches(method.getMethodType());
+    }
+
+    default boolean matches(@Nullable J.MemberReference memberReference) {
+        if (memberReference == null) {
+            return false;
+        }
+        return matches(memberReference.getMethodType());
+    }
+
+    /**
+     * Whether the method invocation or constructor matches the criteria of this matcher.
+     *
+     * @param maybeMethod Any {@link Expression} that might be a method invocation or constructor.
+     * @return True if the invocation or constructor matches the criteria of this matcher.
+     */
+    default boolean matches(@Nullable Expression maybeMethod) {
+        return (maybeMethod instanceof J.MethodInvocation && matches((J.MethodInvocation) maybeMethod)) ||
+               (maybeMethod instanceof J.NewClass && matches((J.NewClass) maybeMethod)) ||
+               (maybeMethod instanceof J.MemberReference && matches((J.MemberReference) maybeMethod));
```

Comment:
I could likely prune this down to the following:
```suggestion
        return maybeMethod instanceof MethodCall && matches(((MethodCall) maybeMethod).getMethodType());
```